src/odin_devices/si5324.py

```diff
@@ -0,0 +1,539 @@
+
+        if verify:
+            verify_value = self.get_register_field(field)
+            logger.debug("Verifying value written ({:b}) against re-read: {:b}".format(value,verify_value))
+            if verify_value != value:
+                raise I2CException(
+                        "Value {} was not successfully written to Field {}".format(value, field))
+
+        if (field.register in SI5324._ICAL_sensitive_registers):
+            logger.info("Register {} requires iCAL run".format(field.register))
+            self.iCAL_required = True
+
+
+    def get_register_field(self, field):
+        """
+        Read only the field-specific bits from the relevant register
+
+        :param field: _Field instance holding relevant register and location of field bits
+        """
+        logger.debug("Getting field starting at bit {}, length {} from register {}".format(field.startbit,field.length,field.register))
+
+        raw_register_value = self.readU8(field.register)
+        logger.debug("Raw value: {0:b}".format(raw_register_value))
+
+        # remove high bits
+        value = raw_register_value & (0xFF >> (7-field.startbit))
+        logger.debug("high bits removed: {0:b}".format(value))
+
+        # shift value to position 0
+        value = value >> field.get_endbit()
+        logger.debug("Low bits removed: {0:b}".format(value))
+        return value
+
+
+    """
+    Register Map File Functions
+    """
+    def apply_register_map (self, mapfile_location, verify = True):
+        """
+        Write configuration from a register map generated with DSPLLsim.
+        Since the map is register rather than value-based, there is no need to make use
+        of the _Field access functions.
+
+        :param mapfile_location: location of register map file to be read
+        :param verify: Boolean. If true, read registers back to verify they are written correctly.
+        """
+        f = open(mapfile_location, 'r')
+
+        for line in f.readlines():
+            # The register map starts after general information is printed preceded by '#'
+            if line[0] != '#':
+                # Extract register-value pairing from register map
+                register, value = line.split(',')
+                register = int(register)
+                value = int(value[1:3],16) # Value is in hex
+
+                if register == 136 and (value & 0x40):
+                    logger.info("Ignoring write to iCAL, will be applied next")
+                    continue
+
+                # Write register value
+                logger.info("Writing register {} with value {:02X}".format(register,value))
+                self.write8(register, value)
+
+                if verify:
+                    verify_value = self.readU8(register)
+                    logger.debug("Verifying value written ({:b}) against re-read: {:b}".format(value,verify_value))
+                    if verify_value != value:
+                       raise I2CException(
```

Comment:
If an exception is raised while the file is open, it may not get closed properly. It's good practice to use a context manager to do this, since the file is automatically closed when the context exits, for instance:
```
with open(mapfile_location, 'r') as f:
   ... do lots of things ...
  if verify_value != value:
      raise I2CException(....)

etc
```
The other advantage is that you don't have to explicitly call `close()` on the file as the context manager does that for you.